### PR TITLE
add integration tests for upgrade to smartbft

### PR DESCRIPTION
**Pull request by shivdeep-singh-ibm**
Signed-off-by: Shivdeep Singh <Shivdeep.Singh@ibm.com>

#### Type of change

- Test update

#### Description

Add integration tests for upgrade from `etcdraft` to `smartbft`.

### CODE REVIEW

File: orderer/consensus/etcdraft/chain.go [hunk 1]
```diff
@@ -1235,6 +1237,39 @@ func pemToDER(pemBytes []byte, id uint64, certType string, logger *flogging.Fabr
...
+
+	nextOrdererConfig, ok := bundle.OrdererConfig()
+	if ok {
+		if nextOrdererConfig.Capabilities().ConsensusTypeMigration() {
+			cfm := c.newConfigMetadata(block)
+			if cfm == nil {
+				return false
+			}
+
+			_, err := ComputeMembershipChanges(c.opts.BlockMetadata, c.opts.Consenters, cfm.Consenters)
```
Comment (yacovm) [inline orderer/consensus/etcdraft/chain.go:1261]: why do you compute membership changes? and then if the error is not nil, we continue? How do you know what is the reason it errors? 

Why not just look at the consensus type and at the blank metadata? why do you need `ComputeMembershipChanges` ?

Comment (shivdeep-singh-ibm) [reply orderer/consensus/etcdraft/chain.go]: I wanted to use `ComputeMemberShip` to detect empty ConfigMetaData. This approach is buggy, So I have changed it now. I will do byte comparison of marshalled `ConfigMetaData` bytes with marshalled empty `ConfigMetaData` structure

File: integration/raft/migration_test.go [hunk 2]
```diff
@@ -869,3 +870,237 @@ func createDeliverEnvelope(n *nwo.Network, signer *nwo.SigningIdentity, blkNum u
...
+						config_ := nwo.GetConfig(network, peer, orderer1, channel)
+						consensusTypeValue_ := extractOrdererConsensusType(config_)
+						oldMetadata := consensusTypeValue_.Metadata
+						return oldMetadata
+					}
+
+					oldRaftMetadata := getRaftMetaDataFromConfig()
+
+					current, updated := prepareTransition(network, peer, orderer1, channel,
+						oldConsensusType, protosorderer.ConsensusType_STATE_NORMAL,
```
Comment (yacovm) [inline integration/raft/migration_test.go:1012]: are we doing both config changes in the same transaction?

1. Putting the channel to maintenance
2. Changing the consensus type

? 

We should first move to maintenance and then change the consensus type.

File: integration/raft/migration_test.go [hunk 3]
```diff
@@ -869,3 +870,237 @@ func createDeliverEnvelope(n *nwo.Network, signer *nwo.SigningIdentity, blkNum u
...
+
+					nwo.UpdateOrdererConfig(network,
+						orderer1,
+						channel,
+						current,
+						updated,
+						peer,
+						orderer1, orderer2, orderer3)
+
+					By("2) Verify: Normal TX's on standard channel are blocked")
```
Comment (yacovm) [inline integration/raft/migration_test.go:1050]: you should verify it *before* sending the update that changes consensus type

File: integration/raft/migration_test.go [hunk 4]
```diff
@@ -869,3 +870,237 @@ func createDeliverEnvelope(n *nwo.Network, signer *nwo.SigningIdentity, blkNum u
...
+						current,
+						updated,
+						peer,
+						orderer1, orderer2, orderer3)
+
+					By("3) Verify: standard channel config changed")
+					std2BlockNum := nwo.CurrentConfigBlockNumber(network, peer, orderer1, channel)
+					Expect(std2BlockNum).To(Equal(std2EntryBlockNum + 1))
+
+					By("3) Verify: delivery request from peer is blocked")
```
Comment (yacovm) [inline integration/raft/migration_test.go:1075]: you need to verify it *before* changing the consensus type

File: orderer/consensus/etcdraft/chain.go [hunk 5]
```diff
@@ -1235,6 +1238,37 @@ func pemToDER(pemBytes []byte, id uint64, certType string, logger *flogging.Fabr
...
+	if ok {
+		if nextOrdererConfig.Capabilities().ConsensusTypeMigration() {
+			cfm := c.newConfigMetadata(block)
+			if cfm == nil {
+				return false
+			}
+			emptyMetaData := &etcdraft.ConfigMetadata{}
+			cfmBytes := protoutil.MarshalOrPanic(cfm)
+			emptyMetaDataBytes := protoutil.MarshalOrPanic(emptyMetaData)
+			if bytes.Equal(cfmBytes, emptyMetaDataBytes) && nextOrdererConfig.ConsensusType() == "smartbft" {
```
Comment (yacovm) [inline orderer/consensus/etcdraft/chain.go:1264]: You are checking here that the metadata is empty, but, this is post consensus. 

We need a check that ensures the validity of this config change when the config update is received *before* consensus.

Comment (shivdeep-singh-ibm) [reply orderer/consensus/etcdraft/chain.go:1264]: - Should we verify signatures on the config before consensus to ensure validity?
- What action is to be taken if validation fails?

Comment (yacovm) [reply orderer/consensus/etcdraft/chain.go:1264]: You already verify the signatures in the broadcast filter. 
If the validation fails you can reject the transaction.

File: orderer/consensus/etcdraft/chain.go [hunk 6]
```diff
@@ -1363,6 +1411,45 @@ func (c *Chain) newConfigMetadata(block *common.Block) *etcdraft.ConfigMetadata
...
+// It returns (true, nil) in the case of a valid config for transition to smartbft and (true, error) in case
+// the new config has a consensus type change to smartbft, but has invalid metadata.
+// It returns (false, nil) in case the new config doesn't change consensus type to smartbft.
+func (c *Chain) validateTransitionToSmartbft(oldOrdererConfig, newOrdererConfig channelconfig.Orderer) (bool, error) {
+	var transitionTobft bool
+	oldCapabilities := oldOrdererConfig.Capabilities()
+	if oldCapabilities == nil {
+		return transitionTobft, nil
+	}
+	if oldCapabilities.ConsensusTypeMigration() {
```
Comment (yacovm) [inline orderer/consensus/etcdraft/chain.go:1430]: you also need to check ConsensusTypeBFT() capability

File: orderer/consensus/etcdraft/chain.go [hunk 7]
```diff
@@ -1363,6 +1411,45 @@ func (c *Chain) newConfigMetadata(block *common.Block) *etcdraft.ConfigMetadata
...
+		ordererConfig := oldOrdererConfig
+		nextOrdererConfig := newOrdererConfig
+		transitionTobft = nextOrdererConfig.ConsensusType() == "smartbft" && ordererConfig.ConsensusType() == "etcdraft"
+
+		if transitionTobft {
+			updatedMetadata := &etcdraft.ConfigMetadata{}
```
Comment (yacovm) [inline orderer/consensus/etcdraft/chain.go:1436]: how can the next metadata be of type etcdraft?

File: orderer/consensus/etcdraft/chain.go [hunk 8]
```diff
@@ -1235,6 +1238,48 @@ func pemToDER(pemBytes []byte, id uint64, certType string, logger *flogging.Fabr
...
+	nextOrdererConfig, ok := ordererConfigFromBlock(block)
+	if !ok {
+		return false
+	}
+
+	transitionTobft, err := c.validateTransitionToSmartbft(oldOrdererConfig, nextOrdererConfig)
+	if transitionTobft && err == nil {
+		return true
+	}
+	return false
```
Comment (yacovm) [inline orderer/consensus/etcdraft/chain.go:1280]: if the error is not nil, how did this get past the pre-consensus checks?

File: orderer/consensus/etcdraft/chain.go (hunk 6 above)
Comment (shivdeep-singh-ibm) [reply orderer/consensus/etcdraft/chain.go:1430]: sure

File: orderer/consensus/etcdraft/chain.go (hunk 8 above)
Comment (shivdeep-singh-ibm) [reply orderer/consensus/etcdraft/chain.go:1280]: it is a redundant check, will remove it.

**Review by yacovm — APPROVED** [review]

**Pull request merged**